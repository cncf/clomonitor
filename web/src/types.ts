export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
