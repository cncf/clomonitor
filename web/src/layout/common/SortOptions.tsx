import { isNull } from 'lodash';
import { ChangeEvent, useRef } from 'react';

import { SortBy, SortDirection } from '../../types';
import styles from './SortOptions.module.css';

interface Props {
  by: SortBy;
  direction: SortDirection;
  onSortChange: (by: SortBy, direction: SortDirection) => void;
}

interface Option {
  label: string;
  by: SortBy;
  direction: SortDirection;
}

const SORT_OPTS: Option[] = [
  {
    label: 'Alphabetically (A-Z)',
    by: SortBy.Name,
    direction: SortDirection.ASC,
  },
  {
    label: 'Alphabetically (Z-A)',
    by: SortBy.Name,
    direction: SortDirection.DESC,
  },
  {
    label: 'Score (highest first)',
    by: SortBy.Score,
    direction: SortDirection.DESC,
  },
  {
    label: 'Score (lowest first)',
    by: SortBy.Score,
    direction: SortDirection.ASC,
  },
];

const SortOptions = (props: Props) => {
  const selectEl = useRef<HTMLSelectElement>(null);

  const handleChange = (event: ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value.split('_');
    props.onSortChange(value[0] as SortBy, value[1] as SortDirection);
    forceBlur();
  };

  const forceBlur = (): void => {
    if (!isNull(selectEl) && !isNull(selectEl.current)) {
      selectEl.current.blur();
    }
  };

  return (
    <div className="d-flex flex-nowrap align-items-center me-2 me-md-4">
      <label className="form-label me-2 mb-0">Sort:</label>
      <select
        ref={selectEl}
        className={`form-select form-select-sm rounded-0 ${styles.select}`}
        value={`${props.by}_${props.direction}`}
        onChange={handleChange}
      >
        {SORT_OPTS.map((opt: Option) => (
          <option key={`sort_${opt.label}`} value={`${opt.by}_${opt.direction}`}>
            {opt.label}
          </option>
        ))}
        ;
      </select>
    </div>
  );
};

export default SortOptions;
