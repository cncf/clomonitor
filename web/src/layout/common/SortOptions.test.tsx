import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SortBy, SortDirection } from '../../types';
import SortOptions from './SortOptions';

const mockOnSortChange = jest.fn();

const defaultProps = {
  by: SortBy.Name,
  direction: SortDirection.ASC,
  onSortChange: mockOnSortChange,
};

describe('SortOptions', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<SortOptions {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(<SortOptions {...defaultProps} />);

    expect(screen.getByText('Sort:')).toBeInTheDocument();

    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(screen.getAllByRole('option').length).toBe(4);
    expect(screen.getByRole('option', { name: 'Alphabetically (A-Z)' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Alphabetically (Z-A)' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Score (highest first)' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Score (lowest first)' })).toBeInTheDocument();
    expect((screen.getByRole('option', { name: 'Alphabetically (A-Z)' }) as HTMLOptionElement).selected).toBe(true);

    expect(screen.getByLabelText('Sort options select')).toBeInTheDocument();
    expect(screen.getByLabelText('Sort options select')).toHaveValue('name_asc');
  });

  it('calls onChange to update select', () => {
    render(<SortOptions {...defaultProps} />);

    userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'Score (highest first)' })
    );

    expect(mockOnSortChange).toHaveBeenCalledTimes(1);
    expect(mockOnSortChange).toHaveBeenCalledWith('score', 'desc');
  });
});
