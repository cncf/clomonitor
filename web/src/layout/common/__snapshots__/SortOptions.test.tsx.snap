// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortOptions creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-nowrap align-items-center me-2 me-md-4"
  >
    <label
      class="form-label me-2 mb-0"
    >
      Sort:
    </label>
    <select
      aria-label="Sort options select"
      class="form-select form-select-sm rounded-0 cursorPointer select"
    >
      <option
        value="name_asc"
      >
        Alphabetically (A-Z)
      </option>
      <option
        value="name_desc"
      >
        Alphabetically (Z-A)
      </option>
      <option
        value="score_desc"
      >
        Score (highest first)
      </option>
      <option
        value="score_asc"
      >
        Score (lowest first)
      </option>
      ;
    </select>
  </div>
</DocumentFragment>
`;
